### ISSUE TITLE
Provide reference topic of XML schema for server.xml file

### ISSUE THREAD

**Issue by lauracowen** [OpenLiberty/docs#2408]
(might need to break this out into separate issue per file schema)

XML schemas are the format in which the structure of a valid XML file is defined. So it's not an example XML file but the context of an XSD file.

We don't currently document this in the KC but @NottyCode  suggested it might be useful (for server.xml and other config files).

According to this blog post, it would be useful to people to know that there is an XSD they can use: https://yeti-it.hr/blog?id=1 

I think we can generate the XSD for server.xml by following the instructions in that blog post. And then just provide it in a reference topic.

**Comment by Charlotte-Holt** [comment]
We'll use this issue to create a `server.xml` schema topic. If it's helpful, we can create more issues for other file schemas in the future.

**Comment by brenthdaniel** [comment]
looks good to me

**Comment by lauracowen** [comment]
https://draft-openlibertyio.mybluemix.net/docs/20.0.0.12/reference/xsd-schema.html

**Comment by lauracowen** [comment]
Thanks. I think most of the info is there but needs reordering/editing mostly:

- The intro says moreorless the same thing three times without explaining why you might want to generate the schema. There's also no different tool for generating the XSD when using Maven than when using without (the third sentence implies there is).
- Don't use the phrase "embedded Open Liberty" - it's just "Open Liberty" - it's not embedded in anything. When you use Maven, you're just using a build tool that downloads OL (and creates and starts a server, and deploys the app to the server) for you instead of you manually downloading it from the website.
-  Why would you want to generate the XSD (what's the point of reading this topic)? (see what the guy says in the blog post)
- Can you follow the instructions in the "Generating XSD schema for Open Liberty section" and check that you can generate the server.xsd file, then modify the instructions if not (the sentence "Download the Open Liberty binary files and extract the archive, which can be found in the wlp directory. " doesn't make sense but you should be able to correct it easily if you try it - if you can't follow that first step, no reader will be able to either).
- Don't say "The Open Liberty binary distribution" - just say "Open Liberty".
- Lead with the Maven approach - that's what we're assuming most developers are using.
- Don't talk about where the jar file can be found; instead tell the user to start by (first step) running the Maven package command on their application and the result of that is to generate that directory containing the jar file.
- Then get them to generate the schema.
- Then tell them to add the schema location to their server.xml file.
- Then tell them the result of that - I'm guessing that they can then validate their server.xml file in whichever IDE they're using.
- I'm not even sure you need the non-Maven version then. On reflection, maybe keep it but make it secondary to the Maven approach.
- If you have Maven installed (you should have if you've tried any of the Open Liberty guides), grab a copy of the Getting Started guide's `finish` app from GitHub, run the Maven package command, then follow the instruction. BTW, if you haven't already, it's worth you trying the Getting Started guide anyway. It's not complicated, though you might need some help with getting Maven installed if you haven't already - but that's worth having anyway, and it will be invaluable in helping you understand the developer's perspective.
- Finally, what you have, while not a direct copy, is very much based on that guy's blog post. Therefore, we should acknowledge it in a sentence at the end of the post, including a link to his original post (https://yeti-it.hr/blog?id=1). Something like "This article was inspired by a blog post by Open Liberty user on the YETI IT blog.".

**Comment by lauracowen** [comment]
I should've said that the reason for prioritising the Maven section over the other is that this topic is squarely targeted at developer needing the schema for, say, their IDE (so it validates their edits of server.xml), as described in the original guy's blog post. Developers are assumed to be using Maven (or similar) build tools.

**Comment by ManasiGandhi** [comment]
I worked on Laura's review. Waiting on the changes to show after the builds issue is resolved.

**Comment by Charlotte-Holt** [comment]
@ManasiGandhi Looks like you addressed most of Laura's feedback. Here's a bit more: 

- [ ] Could we maybe add a sentence to the intro that talks about why someone would want to do this? 
- [ ] I wouldn't use gerunds in the subheadings, maybe just "Generate"?
- [ ] Looks like a typo with the backticks in "generate the`./target/liberty/wlp/bin/tools` directory"
- [ ] that contains ws-schemagen.jar file. --> that contains a ws-schemagen.jar file.
- [ ] Commands that are on lines by themselves need to be in code blocks
- [ ] Don't use monospacing in the subheadings

**Comment by ManasiGandhi** [comment]
@Charlotte-Holt Thanks for your review. I worked on your suggestions. 

- [x]  Could we maybe add a sentence to the intro that talks about why someone would want to do this?

- [x] I wouldn't use gerunds in the subheadings, maybe just "Generate"?

- [x] Looks like a typo with the backticks in "generate the./target/liberty/wlp/bin/tools directory"

- [x] that contains ws-schemagen.jar file. --> that contains a ws-schemagen.jar file.

- [x] Commands that are on lines by themselves need to be in code blocks

- [x] Don't use monospacing in the subheadings

**Comment by Charlotte-Holt** [comment]
@ManasiGandhi Thanks! Content review approved. I'd note that I think you need to remove the backticks from around the commands, but you can include that update after someone completes a peer review. 

https://draft-openlibertyio.mybluemix.net/docs/21.0.0.2/reference/xsd-schema.html

**Comment by Rwalls1** [comment]
# Peer review

## Intro/general comments
- [ ] I think you should change “validate your application `server.xml` file” to “validate the `server.xml` file of your application” to make the relationship between the application and the server.xml file clearer.
- [ ] I think you should avoid the use of commands such as “Generate the XML schema” because this syntax indicates that there is a task to complete rather than the intended purpose of providing examples of XML schema.
## Generate XML schema for Open Liberty with the Maven plug-in
- [ ] I think you should change “For maven based projects” to “For maven based projects with Open Liberty” to help clarify why the command would generate that specific directory
- [ ] I think you should provide a link on the first reference to Maven to provide more context about the tool for users. The link could be to an official source such as https://maven.apache.org/what-is-maven.html.
- [ ] When you reference the “Maven package command”, I think you should reference the code syntax to make the description clearer to users when they run the command so I think you should change “Maven package command” to “`mvn package`” command.
- [ ] The phrase “on your application" seems a bit ambiguous and may be a bit confusing without additional context. In the [getting started guide](https://openliberty.io/guides/getting-started.html#building-and-running-the-application), which mentions the `mvn package` command, it states that the command is run from the start directory to make the `.war` file exist in the target directory, which provides a clearer description of what’s happening. I think you should review that guide and revise your current description to align more with description of the command in the guide. 
- [ ] Typo: change "that contains ws-schemagen.jar file” to “that contains the ws-schemagen.jar file”
- [ ] Typo: change “where wlp directory exists” to “where the wlp directory exists”
- [ ] Typo: change “to the same directory the `server.xml` file exists” to “to the same directory that the server.xml file exists”
- [ ] I think you should change “The default location for the `server.xml` file for Open Liberty” to “The default location for the Open Liberty `server.xml` file” to be more concise.
- [ ] I think you should change “ is `src/main/liberty/config`” to “ is the `src/main/liberty/config directory`” to avoid possible confusion about what’s being referred to.
- [ ] Typo: add a period after the “Set the location” sentence
- [ ] I think the use of the phrase “current directory” is a bit ambiguous and could be confusing. Based on the info contained in the blog, I think you could change “current directory” to something like “ location that is set in the command” to make it clear what directory is being referred to. You should make a similar update in the Open Liberty section when current directory is mentioned
## Generate XML schema for Open Liberty
- [ ] Can you verify that “Download the Open Liberty files and extract the archive” is an action that can be followed as Laura mentioned? It seems unclear and may need to be rephrased to avoid confusion.
- [ ] Typo: change “where `wlp` directory exists” to “where the `wlp` directory exists”
- [ ] Typo: change “XML schema server.xsd file” to “the XML schema server.xsd file”
## Add schema location to the server tag in the server.xml file
- [ ] I think you should change the heading to something like, “Add the schema location to the application” to be more concise and avoid duplicating the description in the heading.
- [ ] I think you should change “add the schema location to the server tag” to “add the schema location to the `xsi : noNamespaceSchemaLocation` attribute in the `server` element of the `server.xml` file” to clarify the location of schema.
- [ ] I think you should also add at least another sentence to clarify why the user needs to know this as it currently doesn’t seem completely clear why this is relevant.
- [ ] Change “by Open Liberty user” to “by an Open Liberty user"

**Comment by ManasiGandhi** [comment]
@Rwalls1 I worked on your peer review. Here is a link to the updated draft https://draft-openlibertyio.mybluemix.net/docs/21.0.0.7/reference/xsd-schema.html

# Peer review
## Intro/general comments
- [x] I think you should change “validate your application `server.xml` file” to “validate the `server.xml` file of your application” to make the relationship between the application and the server.xml file clearer.
- [ ] I think you should avoid the use of commands such as “Generate the XML schema” because this syntax indicates that there is a task to complete rather than the intended purpose of providing examples of XML schema.
## Generate XML schema for Open Liberty with the Maven plug-in
- [x] I think you should change “For maven based projects” to “For maven based projects with Open Liberty” to help clarify why the command would generate that specific directory
- [x] I think you should provide a link on the first reference to Maven to provide more context about the tool for users. The link could be to an official source such as https://maven.apache.org/what-is-maven.html.
- [x] When you reference the “Maven package command”, I think you should reference the code syntax to make the description clearer to users when they run the command so I think you should change “Maven package command” to “`mvn package`” command.
- [x] The phrase “on your application" seems a bit ambiguous and may be a bit confusing without additional context. In the [getting started guide](https://openliberty.io/guides/getting-started.html#building-and-running-the-application), which mentions the `mvn package` command, it states that the command is run from the start directory to make the `.war` file exist in the target directory, which provides a clearer description of what’s happening. I think you should review that guide and revise your current description to align more with description of the command in the guide.
- [x] Typo: change "that contains ws-schemagen.jar file” to “that contains the ws-schemagen.jar file”
- [x] Typo: change “where wlp directory exists” to “where the wlp directory exists”
- [x] Typo: change “to the same directory the `server.xml` file exists” to “to the same directory that the server.xml file exists”
- [x] I think you should change “The default location for the `server.xml` file for Open Liberty” to “The default location for the Open Liberty `server.xml` file” to be more concise.
- [x] I think you should change “ is `src/main/liberty/config`” to “ is the `src/main/liberty/config directory`” to avoid possible confusion about what’s being referred to.
- [x] Typo: add a period after the “Set the location” sentence
- [x] I think the use of the phrase “current directory” is a bit ambiguous and could be confusing. Based on the info contained in the blog, I think you could change “current directory” to something like “ location that is set in the command” to make it clear what directory is being referred to. You should make a similar update in the Open Liberty section when current directory is mentioned
## Generate XML schema for Open Liberty
- [x] Can you verify that “Download the Open Liberty files and extract the archive” is an action that can be followed as Laura mentioned? It seems unclear and may need to be rephrased to avoid confusion. (doc tested by me)
- [x] Typo: change “where `wlp` directory exists” to “where the `wlp` directory exists”
- [x] Typo: change “XML schema server.xsd file” to “the XML schema server.xsd file”
## Add schema location to the server tag in the server.xml file
- [x] I think you should change the heading to something like, “Add the schema location to the application” to be more concise and avoid duplicating the description in the heading.
- [x] I think you should change “add the schema location to the server tag” to “add the schema location to the `xsi : noNamespaceSchemaLocation` attribute in the `server` element of the `server.xml` file” to clarify the location of schema.
- [x] I think you should also add at least another sentence to clarify why the user needs to know this as it currently doesn’t seem completely clear why this is relevant. (described in the short description)
- [x] Change “by Open Liberty user” to “by an Open Liberty user"

**Comment by dmuelle** [comment]
Hi Manasi- I. was just looking at this draft for a related issue and noticed a couple things:

- there are extra backtics in the first two examples- don't need bacticks in a code block
- the link to the YETI blog goes to the wrong page. It should go to https://yeti-it.hr/blog?id=1
- the shortdesc says: 'Open Liberty includes tools that create an XSD file to validate the application server.xml file so that you have the autocompletion capability when you edit the file."  Although the blog author states that is the reason he did this, I think it is a relatively narrow use case, and the context is missing. He went this route because he didn't want to use Eclipse anymore, but most users wouldn't need to generate an XSD just to get autocomplete functions in their IDE, and even if they did, it would require additional steps that aren't mentioned here. I think a more general statement about the value of an XML schema (validating the structure of your server.xml file) would be more appropriate.

**Comment by dmuelle** [comment]
actually, looking at the OP again- the shortdesc might be ok, but the last heading title "Add schema location to the application" is ambiguous. Which application? I think this is in reference to an IDE that you enable to autocomplete but I'm not sure. The OP says:

> To keep things short, if you want autocompletion while editing OpenLiberty's server.xml file and you are using editors like Apache NetBeans, Intellij IDEA or VS Code or any other tool capable of working with XSD schemas for that matter, you can simply include XSD schema in you server.xml file and the editor will take care of the rest.
> Schema location is set to server.xsd because we saved it in the same directory where server.xml file resides. If you saved it somewhere else, make sure to adjust the noNamespaceSchemaLocation accordingly

Some version of these statements would be helpful I think

**Comment by dmuelle** [comment]
see comments in #4409

**Comment by dmuelle** [comment]
also see this stack overflow article: https://stackoverflow.com/questions/58630532/xsd-schema-for-liberty-server-xml

and this slack thread: https://ibm-cloud.slack.com/archives/C31DW78RG/p1626965670220800

**Comment by dmuelle** [comment]
hold off on this issue, pending work from: https://github.com/OpenLiberty/docs/issues/4550

**Comment by dmuelle** [comment]
shcema generation is now documented in the commands reference
